### PR TITLE
[Issue 172] Add key based batcher

### LINKED ISSUE #172: Support key based batcher for key_shared
**Is your feature request related to a problem? Please describe.**

If users want to enable batch in key_shared subscription type, we need to support `key based batcher` feature.

**Describe the solution you'd like**

Support `key based batcher` in go client, reference to: https://github.com/apache/pulsar/pull/4435/

### PR DESCRIPTION
**Pull request by freeznet**
Fixes #172

### Motivation

Add a new batch message container named `keyBasedBatchContainer` to support batching message in key_shared subscription mode.

### Modifications

- add `BatchBuilder` interface, add `FlushBatches` and `IsMultiBatches` func
- change old `BatchBuilder` struct to `batchContainer`
- add `keyBasedBatchContainer`
- add tests

### Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests for key based batch producer with multiple consumer in KeyShared mode*
  - *Added integration tests for message ordering with key based batch producer and KeyShared consumer*

### CODE REVIEW

**Comment by freeznet** [comment]
rerun failure checks

File: pulsar/consumer_test.go [hunk 1]
```diff
@@ -1712,3 +1712,162 @@ func TestConsumerName(t *testing.T) {
...
+		}
+		assert.Equal(
+			t, fmt.Sprintf("value-%d", receivedMessageIndex%10),
+			string(cm.Payload()),
+		)
+		consumer1.Ack(cm.Message)
+		receivedMessageIndex++
+	}
+
+	// TODO: add OrderingKey support
```
Comment (wolfstudy) [inline pulsar/consumer_test.go:1872]: Thanks @freeznet work for this, can you help to create an issue to track this issue?

File: pulsar/consumer_test.go [hunk 2]
```diff
@@ -1712,3 +1712,162 @@ func TestConsumerName(t *testing.T) {
...
+	assert.Equal(t, len(consumer1Keys)*MsgBatchCount, receivedConsumer1)
+	assert.Equal(t, len(consumer2Keys)*MsgBatchCount, receivedConsumer2)
+
+	fmt.Printf("TestKeyBasedBatchProducerConsumerKeyShared received messages consumer1: %d consumser2: %d\n",
+		receivedConsumer1, receivedConsumer2)
+	assert.Equal(t, 300, receivedConsumer1+receivedConsumer2)
+
+	fmt.Printf("TestKeyBasedBatchProducerConsumerKeyShared received messages keys consumer1: %v consumser2: %v\n",
+		consumer1Keys, consumer2Keys)
+}
```
Comment (wolfstudy) [inline pulsar/consumer_test.go:1808]: In this test case, how do we determine that the order of messages received by the consumer is correct？

**Review by wolfstudy — COMMENTED** [review]
Thanks @freeznet , great changes, just a little comment, please check.

File: pulsar/consumer_test.go (hunk 2 above)
Comment (freeznet) [reply pulsar/consumer_test.go:1808]: @wolfstudy thanks for comment. The intention for this test is to check if `KeyBasedBatchBuilder` works ok (batch message with key works, which different from normal batch builder), the order of messages is test with `TestOrderingOfKeyBasedBatchProducerConsumerKeyShared`. But I can make some changes to add order test in `TestKeyBasedBatchProducerConsumerKeyShared` as well.

File: pulsar/consumer_test.go (hunk 1 above)
Comment (freeznet) [reply pulsar/consumer_test.go:1872]: new issue added #401, also add to comment to track.

**Review by wolfstudy — APPROVED** [review]
LGTM +1

**Comment by wolfstudy** [comment]
@freeznet Please merge master code to fix conflict, thanks?

**Comment by freeznet** [comment]
> @freeznet Please merge master code to fix conflict, thanks?

resolved ;)

**Review by wolfstudy — APPROVED** [review]

**Pull request merged**